poc/vdaf_poplar1.sage

```diff
@@ -11,198 +11,206 @@ from sagelib.vdaf import Vdaf, test_vdaf
+                                            level)
+        bound_msg = bound_msg_part_1 + bound_msg_part_2 if agg_id == 0 else b''
+
+        check_bound = b'' if agg_id == 0 \
+                    else Poplar1.hash2(data_sum, auth_sum, level)
+
+        prep = (b'init', (agg_id,
+                                [check_bound, corr_check, bound_msg],
+                                out_share))
+        return prep
 
     @classmethod
-    def prep_next(Poplar1, prep_state, opt_sketch):
-        (step, level, prep_mem) = prep_state
-        Field = Poplar1.Idpf.current_field(level)
-
-        # Aggregators exchange masked input values (step (3.)
-        # of [BBCGGI21, Appendix C.4]).
-        if step == b'ready' and opt_sketch == None:
-            sketch_share, prep_mem = prep_mem[:3], prep_mem[3:]
-            return ((b'sketch round 1', level, prep_mem),
-                    Field.encode_vec(sketch_share))
-
-        # Aggregators exchange evaluated shares (step (4.)).
-        elif step == b'sketch round 1' and opt_sketch != None:
-            prev_sketch = Field.decode_vec(opt_sketch)
-            if len(prev_sketch) == 0:
-                prev_sketch = Field.zeros(3)
-            elif len(prev_sketch) != 3:
-                raise ERR_INPUT # prep message malformed
-            (A_share, B_share, agg_id), prep_mem = \
-                prep_mem[:3], prep_mem[3:]
-            sketch_share = [
-                agg_id * (prev_sketch[0]^2 \
-                            - prev_sketch[1]
-                            - prev_sketch[2]) \
-                    + A_share * prev_sketch[0] \
-                    + B_share
-            ]
-            return ((b'sketch round 2', level, prep_mem),
-                    Field.encode_vec(sketch_share))
-
-        elif step == b'sketch round 2' and opt_sketch != None:
-            prev_sketch = Field.decode_vec(opt_sketch)
-            if len(prev_sketch) == 0:
-                prev_sketch = Field.zeros(1)
-            elif len(prev_sketch) != 1:
-                raise ERR_INPUT # prep message malformed
-            if prev_sketch[0] != Field(0):
-                raise ERR_VERIFY
-            return prep_mem # Output shares
-
-        raise ERR_INPUT # unexpected input
+    def prep_next(Poplar1, prep, inbound):
+        (status, prep_state) = prep
+        #Send verification messages (1 round only)
+        if status == b'init':
+            (agg_id, (check_bound, corr_check, bound_msg), out_share) \
+                                                                    = prep_state
+            new_prep = (b'finish',(agg_id, check_bound, out_share))
+            msg = corr_check + bound_msg
+            return (new_prep, msg)
+        if status == b'finish':
+            (agg_id, check_bound, out_share) = prep_state
+            l = Poplar1.Idpf.Prg.SEED_SIZE
+            possibilities = []
+            for i in range(Poplar1.BOUND+1):
```

Comment:
I think the upper bound should be strict. That is, the range check should assert that the value is in range [0..k).